### ISSUE TITLE
v16.0.0-next.7 not rendering ngFor during SSR

### ISSUE THREAD

**Issue by goetzrobin** [angular/angular#49762]
### Which @angular/* package(s) are the source of the bug?

core, platform-server

### Is this a regression?

Yes

### Description

I have a project that is using a custom SSR setup (Analog and Nitro) to render my application on the server.

In v15.2.5 it successfully renders the list of notes that are provided by a resolver.
The HTML output from server looks like this:
```html
<!DOCTYPE html>
<html class="dark h-full" lang="en">
   <head>
      <script type="module" src="/@vite/client"></script>
      <meta charset="utf-8">
      <title>SPARTAN - Notes App</title>
      <base href="/">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <link rel="icon" type="image/x-icon" href="/assets/spartan.svg">
      <link rel="stylesheet" href="/src/styles.css">
   </head>
   <body class="h-full bg-zinc-900">
      <analog-trpc-root class="max-w-screen-md mx-auto block h-full bg-zinc-900 text-zinc-50" ng-version="15.2.5" ng-server-context="other">
         <router-outlet></router-outlet>
         <analog-trpc-home class="block h-full p-4">
            <div class="justify-center flex mt-20 mb-8 items-center">
               <h1 class="font-medium italic text-6xl text-[#DD0031] font-bold">SPARTAN</h1>
               <img alt="Spartan Logo" src="/assets/spartan.svg" class="ml-2 block w-32">
            </div>
            <h1 class="mb-4 italic font-medium text-xl text-zinc-300 text-center"><span class="text-[#DD0031]">S</span>upabase - <span class="text-[#DD0031]">P</span>risma - <span class="text-[#DD0031]">A</span>nalog - t<span class="text-[#DD0031]">R</span>PC - <span class="text-[#DD0031]">T</span>ailwind - <span class="text-[#DD0031]">A</span>ngular - <span class="text-[#DD0031]">N</span>x </h1>
            <form novalidate="" class="py-2 flex items-center ng-untouched ng-pristine ng-invalid"><input required="" autocomplete="off" name="newTitle" class="focus-visible:ring-2 focus-visible:ring-red-500 focus-visible:outline-0 block w-full appearance-none rounded-lg px-3 py-2 transition-colors text-base leading-tight md:text-sm bg-black/[.05] dark:bg-zinc-50/10 focus:bg-white dark:focus:bg-dark placeholder:text-zinc-500 dark:placeholder:text-zinc-400 contrast-more:border contrast-more:border-current ng-untouched ng-pristine ng-invalid" ng-reflect-required="" ng-reflect-name="newTitle" ng-reflect-model="" value=""><button class="ml-2 focus-visible:ring-2 focus-visible:ring-zinc-50 focus-visible:outline-0 flex items-center justify-center rounded-lg px-2 py-1.5 text-sm font-bold tracking-tight shadow-xl shadow-red-500/20 bg-[#DD0031] hover:bg-opacity-70 text-zinc-800 hover:text-primary-darker">+ </button></form>
            <div class="mt-4">
               <div class="mb-4 p-4 font-normal border border-zinc-500/40 rounded-md">
                  <div class="flex items-center justify-between">
                     <p class="text-sm text-zinc-400">Apr 7, 2023</p>
                     <button class="!text-xs h-6 !bg-opacity-10 hover:!bg-opacity-50 !text-zinc-50 focus-visible:ring-2 focus-visible:ring-zinc-50 focus-visible:outline-0 flex items-center justify-center rounded-lg px-2 py-1.5 text-sm font-bold tracking-tight shadow-xl shadow-red-500/20 bg-[#DD0031] hover:bg-opacity-70 text-zinc-800 hover:text-primary-darker">x </button>
                  </div>
                  <p class="mb-4">hello world</p>
               </div>
               <!--bindings={
                  "ng-reflect-ng-for-of": "[object Object]"
                  }--><!--bindings={
                  "ng-reflect-ng-if": "false"
                  }-->
            </div>
         </analog-trpc-home>
         <!--container-->
      </analog-trpc-root>
      <script type="module" src="/src/main.ts"></script>
   </body>
</html>
```
The code for this component is here:
https://github.com/goetzrobin/spartan/blob/fix-ssr-15/apps/analog-trpc/src/app/pages/index.page.ts

After upgrading to v16.0.0-next.7 I do not get the same output anymore. The ngFor and ngIf directives are replaced with with  comments.
The full HTML output from server is this:
```html
<!DOCTYPE html>
<html class="dark h-full" lang="en">
   <head>
      <script type="module" src="/@vite/client"></script>
      <meta charset="utf-8">
      <title>SPARTAN - Notes App</title>
      <base href="/">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <link rel="icon" type="image/x-icon" href="/assets/spartan.svg">
      <link rel="stylesheet" href="/src/styles.css">
   </head>
   <body class="h-full bg-zinc-900">
      <analog-trpc-root class="max-w-screen-md mx-auto block h-full bg-zinc-900 text-zinc-50" ng-version="16.0.0-next.7" ng-server-context="other" ngh="1">
         <router-outlet></router-outlet>
         <analog-trpc-home class="block h-full p-4" ngh="0">
            <div class="justify-center flex mt-20 mb-8 items-center">
               <h1 class="font-medium italic text-6xl text-[#DD0031] font-bold">SPARTAN</h1>
               <img alt="Spartan Logo" src="/assets/spartan.svg" class="ml-2 block w-32">
            </div>
            <h1 class="mb-4 italic font-medium text-xl text-zinc-300 text-center"><span class="text-[#DD0031]">S</span>upabase - <span class="text-[#DD0031]">P</span>risma - <span class="text-[#DD0031]">A</span>nalog - t<span class="text-[#DD0031]">R</span>PC - <span class="text-[#DD0031]">T</span>ailwind - <span class="text-[#DD0031]">A</span>ngular - <span class="text-[#DD0031]">N</span>x </h1>
            <form novalidate="" class="py-2 flex items-center"><input required="" autocomplete="off" name="newTitle" class="focus-visible:ring-2 focus-visible:ring-red-500 focus-visible:outline-0 block w-full appearance-none rounded-lg px-3 py-2 transition-colors text-base leading-tight md:text-sm bg-black/[.05] dark:bg-zinc-50/10 focus:bg-white dark:focus:bg-dark placeholder:text-zinc-500 dark:placeholder:text-zinc-400 contrast-more:border contrast-more:border-current" ng-reflect-required="" ng-reflect-name="newTitle"><button class="ml-2 focus-visible:ring-2 focus-visible:ring-zinc-50 focus-visible:outline-0 flex items-center justify-center rounded-lg px-2 py-1.5 text-sm font-bold tracking-tight shadow-xl shadow-red-500/20 bg-[#DD0031] hover:bg-opacity-70 text-zinc-800 hover:text-primary-darker">+ </button></form>
            <div class="mt-4">
               <!--container--><!--container-->
            </div>
         </analog-trpc-home>
         <!--container-->
      </analog-trpc-root>
      <script type="module" src="/src/main.ts"></script>
      <script id="ng-state" type="application/json">{&q;__ɵnghData__&q;:[{&q;t&q;:{&q;32&q;:&q;t0&q;,&q;34&q;:&q;t1&q;},&q;c&q;:{&q;32&q;:[],&q;34&q;:[]}},{&q;c&q;:{&q;0&q;:[{&q;i&q;:&q;c258015230&q;,&q;r&q;:1}]}}]}</script>
   </body>
</html>
```
The code for this component is here: https://github.com/goetzrobin/spartan/tree/fix-ssr-16

## TLDR
Project with custom SSR setup using Nitro (AnalogJs.) Running serve command and accessing component gives different server rendered HTML.
v15.2.0 SSR rendering outputs:
```html
       <div class="mt-4">
               <div class="mb-4 p-4 font-normal border border-zinc-500/40 rounded-md">
                  <div class="flex items-center justify-between">
                     <p class="text-sm text-zinc-400">Apr 7, 2023</p>
                     <button class="!text-xs h-6 !bg-opacity-10 hover:!bg-opacity-50 !text-zinc-50 focus-visible:ring-2 focus-visible:ring-zinc-50 focus-visible:outline-0 flex items-center justify-center rounded-lg px-2 py-1.5 text-sm font-bold tracking-tight shadow-xl shadow-red-500/20 bg-[#DD0031] hover:bg-opacity-70 text-zinc-800 hover:text-primary-darker">x </button>
                  </div>
                  <p class="mb-4">hello world</p>
               </div>
               <!--bindings={
                  "ng-reflect-ng-for-of": "[object Object]"
                  }--><!--bindings={
                  "ng-reflect-ng-if": "false"
                  }-->
            </div>
```
v16.0.0-next.7 SSR rendering outputs:
```html
            <div class="mt-4">
               <!--container--><!--container-->
            </div>
```

### Please provide a link to a minimal reproduction of the bug

Need to create stackblitz

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 15.2.5
Node: 16.16.0
Package Manager: yarn 1.22.19
OS: darwin x64

Angular: 15.2.5
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1502.5
@angular-devkit/build-angular   15.2.5
@angular-devkit/core            15.2.5
@angular-devkit/schematics      15.2.5
@schematics/angular             15.2.5
rxjs                            7.8.0
typescript                      4.9.5

----------------------------------------------------------------------------

Angular CLI: 16.0.0-next.7
Node: 16.16.0
Package Manager: yarn 1.22.19
OS: darwin x64

Angular: 16.0.0-next.7
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1600.0-next.7
@angular-devkit/build-angular   16.0.0-next.7
@angular-devkit/core            16.0.0-next.7
@angular-devkit/schematics      16.0.0-next.7
@ngtools/webpack                16.0.0-next.7
@schematics/angular             16.0.0-next.7
rxjs                            7.8.0
typescript                      4.9.5
```

### Anything else?

I need to spend some time and set up a vanilla Angular Universal stackblitz for both versions. I just wanted to share what I currently have to see if there is anything obvious I am missing.

Thanks already for any help!

**Comment by AndrewKushnir** [comment]
@goetzrobin thanks for reporting the problem.

Could you please also provide a minimal repro in a form of GitHub repository, so that we can investigate this problem?

It looks like what might be happening is that the serialization process on the server in your scenario starts too early, before the data gets loaded and rendered. **For testing purposes**, you can try updating this call:

https://github.com/goetzrobin/spartan/blob/825360fc1c041ba051d9c48749b96d1e0ee25c8a/apps/analog-trpc/src/app/pages/index.page.ts#L88-L91

with a `setTimeout` like it [happens here](https://github.com/angular/angular/blob/main/packages/common/http/src/xhr.ts#L349-L362) (i.e. set timeout outside of the call and clear it once the rpc call is completed). This would help us to better understand whether it's related to ZoneJS, which signals to the serialization process too early.

**Comment by goetzrobin** [comment]
@AndrewKushnir thank you for the quick response! I will give that a try.

I am actually using a resolver to get the initial data from the routes data property here:
https://github.com/goetzrobin/spartan/blob/825360fc1c041ba051d9c48749b96d1e0ee25c8a/apps/analog-trpc/src/app/pages/index.page.ts#L14

The fetchPosts function is only used for subsequent calls on the client!
Does that change anything?

**Comment by AndrewKushnir** [comment]
> I am actually using a resolver to get the initial data from the routes data property here:
https://github.com/goetzrobin/spartan/blob/825360fc1c041ba051d9c48749b96d1e0ee25c8a/apps/analog-trpc/src/app/pages/index.page.ts#L14
>
> The fetchPosts function is only used for subsequent calls on the client!
> Does that change anything?

Thanks for additional info. Yes, in this case the resolver function would be the right place to add the `setTimout` for testing purposes.

**Comment by goetzrobin** [comment]
@AndrewKushnir turns out I didn't update the zonejs dependency to 0.13.0. After doing that I now get the expected result.

I am closing this issue and apologize for wasting your time...

Thanks so much for your help and have a great rest of your day!

**Comment by AndrewKushnir** [comment]
@goetzrobin thanks for the update! No worries, happy to hear that everything worked 👍
